p4-fusion/main.cc

```diff
@@ -63,9 +64,77 @@ int Main(int argc, char** argv)
 		return 1;
 	}
 
-	// Set the signal here because it gets reset after P4API library is initialized.
-	std::signal(SIGINT, SignalHandler);
-	std::signal(SIGTERM, SignalHandler);
+	// Block signals from being handled by the main thread, and all future threads.
+	//
+	// - SIGINT, SIGTERM, SIGHUP: thread pool will be shutdown and std::exit will be called
+	// - SIGUSR1: only sent by main() to tell the signal handler thread to exit
+	sigset_t blockedSignals;
+	sigemptyset(&blockedSignals);
+	for (auto sig : { SIGINT, SIGTERM, SIGHUP, SIGUSR1 })
+	{
+		sigaddset(&blockedSignals, sig);
+	}
+
+	int rc = pthread_sigmask(SIG_BLOCK, &blockedSignals, nullptr);
+	if (rc != 0)
+	{
+		ERR("(signal handler) failed to block signals: (" << errno << ") " << strerror(errno));
+		return 1;
+	}
+
+	// Create the thread pool
+	int networkThreads = arguments.GetNetworkThreads();
+	PRINT("Network Threads: " << networkThreads);
+	PRINT("Creating " << networkThreads << " network threads");
+	ThreadPool pool(networkThreads);
+	SUCCESS("Created " << pool.GetThreadCount() << " threads in thread pool");
+
+	// Spawn a thread to handle signals.
+	// The thread will block and wait for signals to arrive and then shutdown the thread pool, unless it receives SIGUSR1,
+	// in which case it will just exit (since main() is handling the shutdown).
+	//
+	// Using a separate thread for purely signal handling allows us to use non-reentrant functions
+	// (such as std::cout, condition variables, etc.) in the signal handler.
+	auto signalHandlingThread = std::thread(
+	    [&blockedSignals, &pool]()
+	    {
+		    auto shutdown = [&pool]()
+		    {
+			    pool.ShutDown();
```

Comment:
Not all the state inside the `ThreadPool` is protected by a mutex. For example, the `Shutdown` method itself starts mutating state immediately without taking a mutex.

So it's not obvious that calling this method is safe from a different thread. After all, later in this function, there is other code which may be concurrently calling other methods on the `ThreadPool`.

Reply:
You're correct. I saw that the shutdown code was already using atomic / condition variables, but I didn't think to check to see if the existing Shutdown logic was actually threadsafe / sound. After reviewing it, I found some issues and attempted to fix them in [`5ed3404` (#9)](https://github.com/sourcegraph/p4-fusion/pull/9/commits/5ed34043ae8c30c1fa3a00f3ad23ddc3d1cc0935) (see the commit message). PTAL?